Fix and improve `diffQMap`

supersaiyansubtlety opened the pull request:
Moves `setupQMap` and `diffQMap` to a `buildSrc/` plugin and greatly improves `setupQMap`:
  - extracts up-to-date logic to a value source so it can be used for both up-to-date checks and in the task action
  - ensures local QMap changes get discarded when switching QEP branches
  - stashes/pops local QMap changes when QEP's branch hasn't changed
  - discards local QMap changes if a `--discard` flag is passed
  - ensures QMap is on the default branch

supersaiyansubtlety committed:
ensure setupQMap keeps QMap's QEP version updated

OroArmor reviewed (approved):
LGTM! Also idk how i missed the original MR for this, its amazing and very useful. 

One small issue, what if im testing QEP changes with QM? how do i make sure those arent overriden? I think that the `diffQMap` task should only do setup if its not cloned. or maybe a `.gitignore`d file that says dont override or something.

supersaiyansubtlety committed:
move publishToMavenLocal dependency from setupQMap to diffQMap

supersaiyansubtlety committed:
add --no-daemon to diffQMap so gradle daemons don't lock files

supersaiyansubtlety committed:
force setupQMap to reset qmap if jar.didWork
move --no-daemon arg to be first

supersaiyansubtlety committed:
check if qmap remote default branch is different from local qmap
explicitly check if qep branch has changed rather than assuming it has if jar.didWork
stash/pop qmap changes when possible

supersaiyansubtlety committed:
move setupQMap and diffQMap to buildSrc plugin
fix several issues with git commands

supersaiyansubtlety committed:
remove unused import

supersaiyansubtlety commented:
ugh, I need to move some of the git checks back to `upToDateWhen`, converting to draft for now

supersaiyansubtlety committed:
add DiffQMapExtension
extract SetupQMapState and check it in an upToDateWhen check

supersaiyansubtlety committed:
minor tweaks

supersaiyansubtlety committed:
don't re-check if .git exists

supersaiyansubtlety committed:
only write qmap qep version if it needs updating

supersaiyansubtlety committed:
tiny changes

supersaiyansubtlety committed:
make extension  name matche plugin name

supersaiyansubtlety committed:
Merge branch 'master' into fix-diffQMap

eristhea reviewed (approved):
i love gradle magic ❤️

supersaiyansubtlety committed:
fix DEFAULT_BRANCH_IN_SHOW_OUTPUT_PATTERN

supersaiyansubtlety committed:
shallow clone
